tests/unit/developers-dao.test.js

```diff
@@ -0,0 +1,231 @@
+const developersDao = appRoot.require('api/v1/db/oracledb/developers-dao');
+const developersSerializer = appRoot.require('api/v1/serializers/developers-serializer');
+
+chai.should();
+chai.use(chaiExclude);
+chai.use(chaiAsPromised);
+
+describe('Test developers-dao', () => {
+  const fakeId = 'fakeId';
+
+  // most tests can use the same 'getConnection' stub but some need specific ones
+  const standardConnStub = () => {
+    sinon.stub(conn, 'getConnection').resolves({
+      execute: (sql, sqlParams) => {
+        const sqlResults = {
+          multiResults: {
+            rows: [{}, {}],
+            outBinds: {
+              outId: [1],
+            },
+          },
+          singleResult: {
+            rows: [{}],
+            outBinds: {
+              outId: [1],
+            },
+          },
+          emptyResult: { rows: [] },
+        };
+        let result;
+        if (sql.includes('DELETE')) {
+          result = sqlResults.emptyResult;
+        } else if ('developerId' in sqlParams) {
+          result = sqlResults.singleResult;
+        } else {
+          result = sqlResults.multiResults;
+        }
+        return result;
+      },
+      close: () => null,
+    });
+  };
+
+  afterEach(() => sinon.restore());
+
+  it('getDevelopers should return multiResult', () => {
+    standardConnStub();
+    const developersSerializerStub = sinon.stub(developersSerializer, 'serializeDevelopers');
+    developersSerializerStub.returnsArg(0);
+
+    const expectedResult = [{}, {}];
+    const result = developersDao.getDevelopers(fakeId);
+    return result.should
+      .eventually.be.fulfilled
+      .and.deep.equal(expectedResult)
+      .then(() => {
+        sinon.assert.callCount(developersSerializerStub, 1);
+      });
+  });
+
+  it('getDeveloperById should return singleResult', () => {
+    standardConnStub();
+    const developersSerializerStub = sinon.stub(developersSerializer, 'serializeDeveloper');
+    developersSerializerStub.returnsArg(0);
+
+    const fulfilledCases = [
+      { expectedResult: {} },
+    ];
+
+    const fulfilledPromises = [];
+    _.each(fulfilledCases, ({ expectedResult }) => {
+      const result = developersDao.getDeveloperById(
+        fakeId,
+      );
+      fulfilledPromises.push(result.should
+        .eventually.be.fulfilled
+        .and.deep.equal(expectedResult)
+        .then(() => {
+          sinon.assert.callCount(developersSerializerStub, fulfilledCases.length);
+        }));
+    });
+    return Promise.all(fulfilledPromises);
+  });
+
+  it('getDeveloperById should be rejected', () => {
+    const rejectedCases = [
+      { testCase: '[]', error: 'Cannot read property \'length\' of undefined' },
+      { testCase: { rows: [{}, {}] }, error: 'Expect a single object but got multiple results' },
+    ];
+
+    const rejectedPromises = [];
+    _.each(rejectedCases, ({ testCase, error }) => {
+      sinon.stub(conn, 'getConnection').resolves({
```

Comment:
@tsoliangwu0130 pointed out that most of the time, the same stub is used for the connection object but some test cases use a different stub. Could you see if you could refactor this so that only `standardConnStub()` is used? Then you could use `beforeEach()` to run that function automatically for each of your test cases.

Also, there's a few cases you test where the `connection.execute()` method is stubbed to return something like `'[]'`. I don't think this is necessary to test though, since the connection object should never return anything other than an Error or a [result object](https://oracle.github.io/node-oracledb/doc/api.html#resultobject).

Reply:
`standardConnStub()` was originally `beforeEach()`. It was already getting hacky when I set it up to work for normal test cases but it quickly became messy when I wanted to test different outputs from `connection.execute`. I would rather change it to how I stubbed it in `games-dao.test` with `createConnStub`.

Reply:
@Ruefa sure, your method to stub connection in `games-dao.test` is clever and neat. Is there any reason you couldn't use the same approach in other DAO tests?

Reply:
@tsoliangwu0130 No and I did use it in `reviews-dao.test`. I wrote `developers-dao.test` first and based it off of examples in other repositories and started going my own direction after that. I'll refactor `createConnStub()` into a helper class and use it in `developers-dao.test`.